Vertical suggested themes: Display suggested theme during sign…

p-jackson opened the pull request:
Epic: Automattic/zelda-private#108

#### Changes proposed in this Pull Request

* Add support for `suggested_theme` vertical property added in D33507-code. If a vertical has a suggested theme it will be used at site creation
* Add support for `preview_screenshots` vertical property. If a vertical has preview screenshots it will be used in the site mockup
* Opts in to both features by sending the `suggest_theme` parameter to `/verticals`

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test with user where `locale = en`
* In A/B test picker choose "test" group for `verticalSuggestedThemes` test
* Create a site starting from `/start`
  * Choose *Business* site type
    * Choose *Restaurants* vertical
      * You should see the Rockfield theme in the preview
      * Create the site, the new site has the Rockfield theme
    * Choose *any other vertical*
      * You see the regular DOM based preview
      * Create the site, the new site has the Maywood theme
  * Choose *Blog* site type
    * Choose *Restaurants* vertical
      * You see the regular DOM based preview
      * Create the site, the new site has the Maywood theme
* `calypso_signup_site_preview_mockup_clicked` still fires when preview is clicked
* Try it on non-retina somehow. (I'm thinking browserstack might do the trick)

Fixes Automattic/zelda-private#188

andrewserong commented:
@p-jackson apologies if you're still in the middle of this one, but I thought I'd take a quick look at the failing e2e tests. Looks like the test failures are real, if you set a vertical by URL, e.g. https://hash-86559a4684e92e8d3dde483a52306d5055d95b7a.calypso.live/start/user?site_type=blog&vertical=food it creates a WSOD. I thought it could be related to how `isSiteTopicFulfilled` calls `submitSiteVertical`, and the change to `providesDependencies` in `steps-pure.js`?

p-jackson commented:
Thanks @andrewserong I thought I had that going too! But I think it was before I made some other changes. The react error message looks like I've got some infinite loop going on :(
Pulling back into in-progress.

p-jackson commented:
Fixed the above issues in 8506ab4886d2410442ef1c720172406a20025f0c. Turns out there was already unit test coverage for this bug, I just needed to update the mocks so `site-topic` had the `themeSlugWithRepo`.

I've added the concept of `optionalDependencies` which won't block a step from being fulfilled if they're not there, but also won't error if they're missing (which is why I added `themeSlugWithRepo` to the `site-topic` step's dependencies list in the first place.

adamziel commented:
A lot of things are happening in this PR. Can we have a separate PR for `optionalDependencies` change with dedicated tests?

adamziel commented:
That's probably not a big deal, but I just wanted to note that sometimes we get two suggestions: Restaurant (p9v10) and Restaurants. Only the latter will bring up the screenshot preview.

<img width="485" alt="Zrzut ekranu 2019-10-15 o 13 42 07" src="https://user-images.githubusercontent.com/205419/66828565-a5872e00-ef51-11e9-9037-53534f1a28bd.png">

adamziel commented:
Also a discussion prompt: the site title step around no longer updates the site preview with the screenshot approach, which means this would test more than one change. It doesn't seem like a big deal and I think we could get away with leaving the site title step in place for the initial round of testing - but if it doesn't perform as good as we hope it would we could try hiding it in a follow-up test. What do you think @p-jackson @michaeldcain?

adamziel commented:
Also, this PR is good to go without it, but as a follow-up it would be good to figure out how to load these previews faster. On my retina mac it takes anywhere between 2-4s just to load that screenshot - it's a really long time. Some ideas:

* Load lo-fi screenshot first and only then load hi-fi 
* Split that screenshot into few smaller ones and load above-the-fold part first
* Show a better loading indicator like a blinking mockup of a site
* Start loading the screenshot as soon as user hovers over the "Restaurants" item or starts typing something related to Restaurants
* ...?

adamziel commented:
Other than that, it looks really good. I'm not approving it yet because of the flow controller change but I think it's good to go!

p-jackson commented:
> Can we have a separate PR for optionalDependencies change with dedicated tests?

- #36800
- #36805

> Can we make sure the suggestions for Restaurants are not ambiguous? On occasions it suggests Restaraunt (p9v10, spelled with a typo - we should fix that) and Restaurants.

I'd assumed this was intentional! I hate that the name for `p1` is "Restaurants" the plural, why would I say the subject of my business is multiple restaurants??? I have no idea why that's the high performing vertical.

Because I only wanted the ab test to target the highest performing verticals it doesn't show the screenshot for the children of `p1`, maybe that was the wrong decision?

> It doesn't seem like a big deal and I think we could get away with leaving the site title step in place for the initial round of testing

I'd stick with it for this round. There's a number of reasons why the test might not perform as we'd hope this round, it already feels like a bigger than "just one thing" test.

> On my retina mac it takes anywhere between 2-4s just to load that screenshot

Yes this was something I was planning on investigating after this PR. The quickest thing to test might be to lower the jpeg compression a bit and see how bad the text looks. I do like the idea of loading the image "above the fold" first though!

p-jackson commented:
If these tests pass then we're good to go. (I'm not just randomly mashing the tests, honest, I have seen them all individually passing at different stages).

michaeldcain commented:
In general this works really well. Nice work!

> It doesn't seem like a big deal and I think we could get away with leaving the site title step in place for the initial round of testing - but if it doesn't perform as good as we hope it would we could try hiding it in a follow-up test.

I think we're good to leave as-is for this test.

> Also, this PR is good to go without it, but as a follow-up it would be good to figure out how to load these previews faster.

I agree that this isn't a blocker for this PR, given that the test isn't enabled, but there does seem to be a good bit of optimization that can be done on the images. For the JSON previews, we used Photon query args to downsize the images on the fly in `downsize_content_images()`. [The image used on the showcase page](https://i0.wp.com/theme.files.wordpress.com/2019/08/rockfield-homepage.png) is also significantly smaller to start (1.6MB vs 2.4MB). Let me know if you want some help optimizing the images we'll be using.

> I hate that the name for p1 is "Restaurants" the plural, why would I say the subject of my business is multiple restaurants???

I think this is a mistake in the original data source. I agree that we should update it.

> Because I only wanted the ab test to target the highest performing verticals it doesn't show the screenshot for the children of p1, maybe that was the wrong decision?

That's the correct decision for this test. I figured that targeting the "popular" suggestions only would still give us enough data while making the test easier to setup.
